dal/elastic.go

```diff
@@ -2,126 +2,109 @@ package dal
 }
 
-func ElasticDeleteBook(client *elastic.Client, id string) *elastic.DeleteResponse{
+func (e *Elastic) SearchBook(title string, authorName string, priceRangeStr string) ([]models.Book, error) {
 	ctx := context.Background()
-	response, err := client.Delete().Index("books_yuval").Id(id).Do(ctx)
+	query := elastic.NewBoolQuery()
+	if FieldExist(title) {
+		matchQuery1 := elastic.NewMatchQuery("title", title)
+		query.Filter(matchQuery1)
+	}
+	if FieldExist(authorName) {
+		matchQuery2 := elastic.NewMatchQuery("author_name", authorName)
+		query.Filter(matchQuery2)
+	}
+	if FieldExist(priceRangeStr) {
+		priceRange := GetPriceRange(priceRangeStr)
+		if priceRange != nil {
+			rangeQuery := elastic.NewRangeQuery("price").From(priceRange[0]).To(priceRange[1])
+			query.Filter(rangeQuery)
+		}
+	}
+	searchResult, err := (e.Client).Search().Index(INDEX_NAME).Query(query).Size(100).Do(ctx)
 	if err != nil {
-		// Handle error
-		panic(err)
+		return nil, err
 	}
-	return response
+	numHits := searchResult.Hits.TotalHits.Value
+	resultsArr := make([]models.Book, numHits)
+	for i, hit := range searchResult.Hits.Hits {
+		json.Unmarshal(hit.Source, &resultsArr[i])
+	}
+	return resultsArr, nil
 }
 
-func ElasticSearchBook(client *elastic.Client, title string, author_name string, price_range [2]float64, opcode int) *elastic.SearchResult{
+func (e *Elastic) GetStoreInfo() (int64, *float64, error) {
 	ctx := context.Background()
-	query := elastic.NewBoolQuery()
-	matchQuery1 := elastic.NewMatchQuery("title", title)
-	matchQuery2 := elastic.NewMatchQuery("author_name", author_name)
-	rangeQuery := elastic.NewRangeQuery("price").From(price_range[0]).To(price_range[1])
-	var searchQuery *elastic.BoolQuery
-	switch opcode{//opcode specifies which arguments were given to query
-		case 0://no search arguments
-			searchQuery =query.Must()
-		case 1://only price range specified
-			searchQuery = query.Must(rangeQuery)
-		case 2://only title
-			searchQuery = query.Must(matchQuery1)
-		case 3://price and title
-			searchQuery = query.Must(matchQuery1, rangeQuery)
-		case 4://only author
-			searchQuery = query.Must(matchQuery2)
-		case 5://price and author
-			searchQuery = query.Must(matchQuery2, rangeQuery)
-		case 6://title and author
-			searchQuery = query.Must(matchQuery1, matchQuery2)
-		case 7://title author and price
-			searchQuery = query.Must(matchQuery1, matchQuery2, rangeQuery)
+	countService := elastic.NewCountService(e.Client)
+	countResult, err := countService.Index(INDEX_NAME).Do(ctx)
+	if err != nil {
+		return 0, nil, err
 	}
-	searchResult, err := client.Search().Index(INDEXNAME).Query(searchQuery).Size(100).Do(ctx)
+	agg := elastic.NewCardinalityAggregation().Field("author_name.keyword")
+	diffAuthors, err := (e.Client.Search()).Index(INDEX_NAME).Aggregation("diff_authors", agg).Do(ctx)
 	if err != nil {
-		panic(err)
+		return 0, nil, err
 	}
-		return searchResult
+	numDiffAuthors, _ := diffAuthors.Aggregations.Cardinality("diff_authors")
```

Comment:
what if Found is false? 
and why `diff_authors` name? It does not count the number of authors?